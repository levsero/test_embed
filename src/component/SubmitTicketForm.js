/** @jsx React.DOM */

module React from 'react/addons';
module ReactForms from 'react-forms';

import { submitTicketSchema } from 'component/SubmitTicketSchema';
import { i18n }               from 'service/i18n';
require('imports?_=lodash!lodash');

var classSet = React.addons.classSet,
    SubmitTicketFormBody = ReactForms.Form,
    isFailure = ReactForms.validation.isFailure;

var SubmitTicketForm = React.createClass({
  getInitialState() {
    return {
      isValid: false,
      buttonMessage: i18n.t('embeddable_framework.submitTicket.form.submitButton.label.send'),
      isSubmitting: false,
      showBackButton: false
    };
  },

  getDefaultProps() {
    return {
      fullscreen: false
    };
  },

  handleSubmit(e) {
    var form = this.refs.form,
        formValue = form.value(),
        isFormInvalid = isFailure(formValue.validation);

    if (!isFormInvalid) {
      this.setState({
        buttonMessage: i18n.t('embeddable_framework.submitTicket.form.submitButton.label.sending'),
        isSubmitting: true
      });
    }

    this.props.submit(e, {
      isFormInvalid: isFormInvalid,
      value: formValue.value
    });
  },

  handleUpdate(values, isValid) {
    this.setState({isValid: isValid});
  },

  handleBackClick() {
    this.props.onBackClick();
  },

  render() {
    /* jshint quotmark:false */
    var formBody = this.transferPropsTo(
          <SubmitTicketFormBody
            ref='form'
            schema={submitTicketSchema(this.props.customFields)}
            onUpdate={this.handleUpdate}
            component={React.DOM.div} />
        ),
        navigationButtonClasses = classSet({
          'Button Button--nav custom-textColor': true,
          'u-inlineBlock u-posEndS--vert': !this.props.fullscreen,
          'u-posAbsolute u-posStart--vert u-textSizeBaseMobile': this.props.fullscreen,
          'u-isHidden': !this.state.showBackButton
        }),
        buttonClasses = classSet({
          'Button Button--cta Anim-color u-textNoWrap custom-backgroundColor': true,
          'u-pullRight': !this.props.fullscreen,
          'u-sizeFull': this.props.fullscreen
        }),
        titleClasses = classSet({
          'u-textSizeMed u-textBold u-extSizeMed u-textCenter': true,
          'u-posAbsolute u-posCenter': this.state.showBackButton && !this.props.fullscreen,
          'u-posStart--vert': this.state.showBackButton && !this.props.fullscreen,
          'u-marginTM u-textSizeBaseMobile': this.props.fullscreen
        }),
        barClasses = classSet({
          'Form-cta u-cf Container-pullout u-paddingBS': true,
          'Form-cta--bar u-marginBM': !this.props.fullscreen
        });

    return (
      <form
        noValidate
        onSubmit={this.handleSubmit}
        className={'Form u-cf ' + this.props.className}>
        <div className={barClasses}>
          <button
            onClick={this.handleBackClick}
            className={navigationButtonClasses}>
            <i className='Icon Icon--arrow u-textInheritColor' />
            {i18n.t('embeddable_framework.navigation.back')}
          </button>
          <h2 className={titleClasses}>
            {i18n.t('embeddable_framework.submitTicket.form.title')}
          </h2>
        </div>
        {formBody}
        {this.props.children}
        <input
          type='submit'
          value={this.state.buttonMessage}
          ref='submitButton'
          disabled={!this.state.isValid || this.state.isSubmitting}
          className={buttonClasses}
        />
      </form>
    );
  }
});

export { SubmitTicketForm };
