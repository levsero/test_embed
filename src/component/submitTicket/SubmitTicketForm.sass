@import '../../styles/_vars.scss';

.form {
  composes: u-cf from '../../styles/utils.scss';
  composes: Form from '../../styles/components/Form.scss';
}

.hidden {
  composes: u-isHidden from '../../styles/utils.scss';
}

.ticketFormTitle {
  composes: ticketFormsListTitle from './SubmitTicket.sass';
  composes: u-marginBM from '../../styles/utils.scss';
}

.ticketFormContainer {
  composes: u-paddingTN from '../../styles/utils.scss';
}
