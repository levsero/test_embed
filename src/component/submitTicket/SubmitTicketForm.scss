@import '../../styles/_vars.scss';

.form {
  composes: u-cf from global;
  margin: 0;
}

.container {
  height: auto;
}

.containerMobile {
  font-size: $font-size-mobile;
}

.hidden {
  composes: u-isHidden from global;
}

.ticketFormTitle {
  composes: ticketFormsListTitle from './SubmitTicket.scss';
  composes: u-marginBM from global;
}

.ticketFormTitleMobile {
  font-size: $font-size-mobile;
}

.ticketFormContainer {
  composes: u-paddingTN from global;
}

.formWrapper {
  margin-bottom: #{15/$font-size}rem;
}

.buttonGroup {
  margin-top: #{15/$font-size}rem;
  margin-bottom: #{15/$font-size}rem;
}

.button {
  max-width: #{120/$font-size}rem !important;
  padding-right: #{10/$font-size}rem !important;
  padding-left: #{10/$font-size}rem !important;
}
