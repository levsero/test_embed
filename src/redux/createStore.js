import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import onStateChange from 'redux-on-state-change';
import createLogger from 'redux-logger';

import { store } from 'service/persistence';
import { getEnvironment } from 'src/util/utils';

import reducer from 'src/redux/modules/reducer';
import onStateChangeFn from 'src/redux/middleware/onStateChange';

import { sendBlips } from 'src/redux/middleware/blip';

function loggerTitleFormatter(storeName) {
  return (action, time, took) => [`%c[${storeName}] ${String(action.type)}`, `%c@ ${time}`].join(' ');
}

export default function(storeName = 'web_widget') {
  const enableLogging = __DEV__ || getEnvironment() === 'staging' || store.get('debug');
  const logger = createLogger({ titleFormatter: loggerTitleFormatter(storeName) });
  const devToolsExtension = window.parent.__REDUX_DEVTOOLS_EXTENSION__
    && window.parent.__REDUX_DEVTOOLS_EXTENSION__({ name: storeName });
  const middlewares = [
    thunk,
    onStateChange(onStateChangeFn),
    sendBlips
  ];
  let storeEnhancers;

  if (enableLogging) {
    storeEnhancers = !!devToolsExtension
      ? [applyMiddleware(...middlewares), devToolsExtension]
      : [applyMiddleware(...middlewares, logger)];
  } else {
    storeEnhancers = [applyMiddleware(...middlewares)];
  }

  return compose(...storeEnhancers)(createStore)(reducer);
}
