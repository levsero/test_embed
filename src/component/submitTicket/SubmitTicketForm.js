import React, { Component, PropTypes } from 'react';
import ReactDOM from 'react-dom';
import _ from 'lodash';

import { locals as styles } from './SubmitTicketForm.sass';

import { AttachmentList } from 'component/attachment/AttachmentList';
import { Button } from 'component/button/Button';
import { ButtonSecondary } from 'component/button/ButtonSecondary';
import { ButtonGroup } from 'component/button/ButtonGroup';
import { Field } from 'component/field/Field';
import { ScrollContainer } from 'component/container/ScrollContainer';
import { i18n } from 'service/i18n';
import { getCustomFields } from 'utility/fields';

const sendButtonMessageString = 'embeddable_framework.submitTicket.form.submitButton.label.send';
const sendingButtonMessageString = 'embeddable_framework.submitTicket.form.submitButton.label.sending';
const cancelButtonMessageString = 'embeddable_framework.submitTicket.form.cancelButton.label.cancel';
const initialState = {
  attachments: [],
  buttonMessage: sendButtonMessageString,
  cancelButtonMessage: cancelButtonMessageString,
  isRTL: i18n.isRTL(),
  isSubmitting: false,
  isValid: false,
  shouldRemoveForm: false,
  showErrorMessage: false
};

export class SubmitTicketForm extends Component {
  static propTypes = {
    attachmentsEnabled: PropTypes.bool,
    attachmentSender: PropTypes.func.isRequired,
    children: PropTypes.element.isRequired,
    customFields: PropTypes.array,
    disableAutoComplete: PropTypes.bool,
    expanded: PropTypes.bool,
    formState: PropTypes.object,
    formTitleKey: PropTypes.string.isRequired,
    fullscreen: PropTypes.bool,
    hide: PropTypes.bool,
    maxFileCount: PropTypes.number,
    maxFileSize: PropTypes.number,
    onCancel: PropTypes.func,
    previewEnabled: PropTypes.bool,
    setFormState: PropTypes.func,
    subjectEnabled: PropTypes.bool,
    submit: PropTypes.func.isRequired
  };

  static defaultProps = {
    attachmentsEnabled: false,
    children: <span />,
    customFields: [],
    disableAutoComplete: false,
    expanded: false,
    formState: {},
    fullscreen: false,
    hide: false,
    maxFileCount: 5,
    maxFileSize: 5 * 1024 * 1024,
    onCancel: () => {},
    previewEnabled: false,
    setFormState: () => {},
    subjectEnabled: false
  };

  constructor(props, context) {
    super(props, context);

    this.state = _.extend({}, initialState, {
      isValid: props.previewEnabled,
      ticketForm: null,
      ticketFormFields: []
    });
  }

  componentDidMount = () => {
    const customFields = getCustomFields(this.props.customFields, this.props.formState, this.props.disableAutoComplete);
    const showShadow = customFields.fields.length > 0 || this.props.attachmentsEnabled;

    this.refs.scrollContainer.setScrollShadowVisible(showShadow);
  }

  componentDidUpdate = () => {
    if (this.refs.formWrapper && this.props.formState && this.state.shouldRemoveForm) {
      const form = ReactDOM.findDOMNode(this.refs.form);

      _.forEach(form.elements, function(field) {
        if (field.type === 'submit') {
          return;
        }

        if (this.props.formState[field.name]) {
          if (field.type === 'checkbox') {
            // Based on formState set checked property
            field.checked = !!this.props.formState[field.name];
          } else {
            field.value = this.props.formState[field.name];
          }
        } else {
          // If clearing form after submit we need to make sure
          // formState clears out undefined values
          field.value = '';

          // Don't need to check for type here as non checkbox inputs
          // will ignore this property.
          field.checked = false;
        }
      }, this);
    }
  }

  resetTicketFormVisibility = () => {
    // if the user closes and reopens, we need to
    // re-render the search field
    this.setState({ shouldRemoveForm: false });
  }

  focusField = () => {
    const form = ReactDOM.findDOMNode(this.refs.form);

    // Focus on the first empty text or textarea
    const element = _.find(form.querySelectorAll('input, textarea'), function(input) {
      return input.value === '' && _.includes(['text', 'textarea', 'email'], input.type);
    });

    if (element) element.focus();
  }

  hideVirtualKeyboard = () => {
    this.setState({ shouldRemoveForm: true });
  }

  failedToSubmit = () => {
    this.setState({
      isSubmitting: false,
      buttonMessage: sendButtonMessageString
    });

    this.refs.scrollContainer.scrollToBottom();
  }

  handleSubmit = (e) => {
    if (this.props.previewEnabled) {
      e.preventDefault();
      return;
    }

    const isFormValid = this.state.isValid;

    if (isFormValid) {
      this.setState({
        buttonMessage: sendingButtonMessageString,
        isSubmitting: true
      });
    }

    this.props.submit(e, {
      isFormValid: isFormValid,
      value: this.getFormState()
    });
  }

  openAttachment = () => {
    this.setState({ showAttachmentForm: true });
  }

  getFormState = () => {
    const form = ReactDOM.findDOMNode(this.refs.form);

    return _.chain(form.elements)
      .reject((field) => field.type === 'submit')
      .reduce((result, field) => {
        result[field.name] = (field.type === 'checkbox')
                           ? field.checked ? 1 : 0
                           : field.value;

        return result;
      },
      {}).value();
  }

  filterPrefill = (fields, rawPrefillData) => {
    const permittedFields = ['text', 'textarea', 'description', 'integer', 'decimal', 'subject'];

    // Cleans data by removing fields we do not want to enable pre-fill on
    return _.filter(rawPrefillData.fields, (prefillTicketField) => {
      // Intentional non-strict matching between integer and string ids
      const matchingField = _.find(fields, (field) => field.id == prefillTicketField.id); // eslint-disable-line

      return matchingField
      ? _.includes(permittedFields, matchingField.type)
      : false;
    });
  }

  prefillFormState = (fields, rawPrefillData = {}) => {
    const isPrefillValid = rawPrefillData.fields
                           && Array.isArray(rawPrefillData.fields)
                           && rawPrefillData.fields.length !== 0;

    if (!isPrefillValid) return;

    const fieldsPrefill = this.filterPrefill(fields, rawPrefillData);

    // Check if pre-fill is still valid after processing
    if (fieldsPrefill.length === 0) return;

    let formState = this.getFormState();
    const currentLocale = i18n.getLocale();

    fieldsPrefill.forEach((field) => {
      formState[field.id] = field.prefill[`${currentLocale}`] || field.prefill['*'] || '';
    });

    this.props.setFormState(formState);
  }

  updateTicketForm = (form, fields, prefill) => {
    this.setState({
      ticketForm: form,
      ticketFormFields: fields
    }, () => this.prefillFormState(fields, prefill));
  }

  updateForm = () => {
    const form = ReactDOM.findDOMNode(this.refs.form);
    const attachmentsReady = this.props.attachmentsEnabled
                           ? this.refs.attachments.attachmentsReady()
                           : true;

    this.props.setFormState(this.getFormState());
    this.setState({
      isValid: form.checkValidity() && attachmentsReady
    });
  }

  resetState = () => {
    this.setState(initialState);
  }

  handleOnDrop = (files) => {
    this.refs.attachments.handleOnDrop(files);

    setTimeout(() => this.refs.scrollContainer.scrollToBottom(), 0);
  }

  handleAttachmentsError = () => {
    setTimeout(() => this.refs.scrollContainer.scrollToBottom(), 0);
  }

  clear = () => {
    const formData = this.props.formState;
    const form = this.refs.form;

    _.forEach(form.elements, (field) => {
      if (this.props.formState[field.name] && field.type === 'checkbox') {
        field.checked = false;
      }
    });

    if (this.props.attachmentsEnabled) {
      this.refs.attachments.clear();
    }

    this.setState(initialState);
    this.props.setFormState({
      name: formData.name,
      email: formData.email
    });
  }

  renderSubjectField = () => {
    const placeholder = i18n.t('embeddable_framework.submitTicket.field.subject.label', {
      fallback: 'Subject'
    });

    return !this.props.subjectEnabled
         ? null
         : <Field
            placeholder={placeholder}
            value={this.props.formState.subject}
            disableAutoComplete={this.props.disableAutoComplete}
            name='subject'
            disabled={this.props.previewEnabled} />;
  }

  renderEmailField = () => {
    return (
      <Field
        placeholder={i18n.t('embeddable_framework.form.field.email.label')}
        type='email'
        required={true}
        disableAutoComplete={this.props.disableAutoComplete}
        pattern="[a-zA-Z0-9!#$%&'*+/=?^_`{|}~\-`']+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~\-`']+)*@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?" // eslint-disable-line
        value={this.props.formState.email}
        name='email'
        disabled={this.props.previewEnabled} />
    );
  }

  renderNameField = () => {
    return (
      <Field
        placeholder={i18n.t('embeddable_framework.submitTicket.field.name.label')}
        value={this.props.formState.name}
        name='name'
        disableAutoComplete={this.props.disableAutoComplete}
        disabled={this.props.previewEnabled} />
    );
  }

  renderDescriptionField = () => {
    return (
      <Field
        placeholder={i18n.t('embeddable_framework.submitTicket.field.description.label')}
        required={true}
        value={this.props.formState.description}
        name='description'
        disableAutoComplete={this.props.disableAutoComplete}
        input={<textarea rows='5' />}
        disabled={this.props.previewEnabled} />
    );
  }

  renderTicketFormBody = () => {
    const { ticketForm, ticketFormFields } = this.state;
    const formTicketFields = _.filter(ticketFormFields, (field) => {
      return ticketForm.ticket_field_ids.indexOf(field.id) > -1;
    });
    const ticketFieldsElem = getCustomFields(formTicketFields, this.props.formState, this.props.disableAutoComplete);
    const titleMobileClasses = this.props.fullscreen ? styles.ticketFormTitleMobile : '';

    ticketFieldsElem.allFields.unshift([this.renderNameField(), this.renderEmailField()]);

    return (
      <div ref='formWrapper'>
        <div className={`${styles.ticketFormTitle} ${titleMobileClasses}`}>
          {ticketForm.display_name}
        </div>
        {ticketFieldsElem.allFields}
        {this.props.children}
      </div>
    );
  }

  renderFormBody = () => {
    const customFields = getCustomFields(this.props.customFields, this.props.formState, this.props.disableAutoComplete);

    return (
      <div ref='formWrapper'>
        {this.renderNameField()}
        {this.renderEmailField()}
        {customFields.fields}
        {this.renderSubjectField()}
        {this.renderDescriptionField()}
        {customFields.checkboxes}
        {this.props.children}
      </div>
    );
  }

  renderCancelButton = () => {
    const { onCancel, fullscreen } = this.props;

    return (
      <ButtonSecondary
        label={i18n.t(this.state.cancelButtonMessage)}
        onClick={onCancel}
        fullscreen={fullscreen} />
    );
  }

  renderAttachments = () => {
    const { attachmentSender, fullscreen } = this.props;

    return (
      <AttachmentList
        ref="attachments"
        attachmentSender={attachmentSender}
        updateForm={this.updateForm}
        maxFileCount={this.props.maxFileCount}
        maxFileSize={this.props.maxFileSize}
        fullscreen={fullscreen}
        handleAttachmentsError={this.handleAttachmentsError} />
    );
  }

  render = () => {
    const { attachmentsEnabled, fullscreen, formTitleKey, hide } = this.props;

    const form = this.state.ticketForm ? this.renderTicketFormBody() : this.renderFormBody();
    const formBody = this.state.shouldRemoveForm ? null : form;
    const buttonCancel = fullscreen ? null : this.renderCancelButton();
    const attachments = attachmentsEnabled ? this.renderAttachments() : null;
    const hiddenClass = hide ? styles.hidden : '';
    const containerClasses = this.state.ticketForm ? styles.ticketFormContainer : '';

    return (
      <form
        noValidate={true}
        onSubmit={this.handleSubmit}
        onChange={this.updateForm}
        ref='form'
        className={`${styles.form} ${hiddenClass}`}>
        <ScrollContainer
          ref='scrollContainer'
          title={i18n.t(`embeddable_framework.submitTicket.form.title.${formTitleKey}`)}
          contentExpanded={this.props.expanded}
          containerClasses={`${styles.container} ${containerClasses}`}
          footerContent={
            <ButtonGroup rtl={i18n.isRTL()}>
              {buttonCancel}
              <Button
                fullscreen={fullscreen}
                label={i18n.t(this.state.buttonMessage)}
                disabled={!this.state.isValid || this.state.isSubmitting}
                type='submit' />
            </ButtonGroup>
          }
          fullscreen={fullscreen}>
          {formBody}
          {attachments}
        </ScrollContainer>
      </form>
    );
  }
}
