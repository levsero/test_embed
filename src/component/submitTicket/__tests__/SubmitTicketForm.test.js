import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import { ThemeProvider } from 'styled-components'

import { IdManager } from '@zendeskgarden/react-selection'
import { i18n } from 'service/i18n'

import { SubmitTicketForm } from '../SubmitTicketForm'
import createStore from 'src/redux/createStore'
import { Provider } from 'react-redux'

jest.useFakeTimers()

const renderSubmitTicketForm = (props, renderer) => {
  const defaultProps = {
    attachmentSender: noop,
    submit: noop,
    formTitle: 'message',
    nameFieldRequired: false,
    nameFieldEnabled: true
  }
  const mergedProps = { ...defaultProps, ...props }
  const component = (
    <ThemeProvider theme={{}}>
      <Provider store={createStore()}>
        <SubmitTicketForm {...mergedProps} />
      </Provider>
    </ThemeProvider>
  )
  IdManager.setIdCounter(0)
  return renderer ? renderer(component) : render(component)
}

test('renders the expected components', () => {
  const { container } = renderSubmitTicketForm()

  expect(container).toMatchSnapshot()
})

test('renders the form title', () => {
  const { queryByText } = renderSubmitTicketForm({ formTitle: 'hello' })

  expect(queryByText('hello')).toBeInTheDocument()
})

describe('name field', () => {
  it('renders with correct props when it is required', () => {
    const { getByLabelText } = renderSubmitTicketForm({ nameFieldRequired: true })

    expect(getByLabelText('Your name').required).toEqual(true)
  })

  it('renders with the expected prop and label when it is not required', () => {
    const { getByLabelText } = renderSubmitTicketForm({ nameFieldRequired: false })

    expect(getByLabelText('Your name (optional)').required).toBeFalsy()
  })

  it('does not render the name field with it is not enabled', () => {
    const { container } = renderSubmitTicketForm({ nameFieldEnabled: false })

    expect(container.querySelector('input[name="name"]')).not.toBeInTheDocument()
  })
})

describe('email field', () => {
  it('renders with correct props', () => {
    const { getByLabelText } = renderSubmitTicketForm()

    expect(getByLabelText('Email address').required).toEqual(true)
  })
})

describe('description field', () => {
  it('renders with correct props', () => {
    const { getByLabelText } = renderSubmitTicketForm({
      activeTicketForm: null
    })

    expect(getByLabelText('How can we help you?').required).toEqual(true)
  })
})

describe('subject field', () => {
  it('renders with expected label', () => {
    const { queryByLabelText } = renderSubmitTicketForm({
      activeTicketForm: null,
      subjectEnabled: true
    })

    expect(queryByLabelText('Subject (optional)')).toBeInTheDocument()
  })
})

describe('footer', () => {
  it('renders scrollContainer without footer shadow when attachments and ticket fields are false', () => {
    const { container } = renderSubmitTicketForm()

    expect(container.querySelector('footer').className).not.toContain('footerShadow')
  })

  it('renders scrollContainer with footer when attachments are true', () => {
    const { container } = renderSubmitTicketForm({
      attachmentsEnabled: true
    })

    expect(container.querySelector('footer').className).toContain('footerShadow')
  })

  it('renders scrollContainer with footer when ticket fields are true', () => {
    const { container } = renderSubmitTicketForm({
      ticketFields: [1, 2, 3]
    })

    expect(container.querySelector('footer').className).toContain('footerShadow')
  })
})

describe('ticket forms', () => {
  /* eslint-disable camelcase */
  const activeTicketForm = {
    id: 1,
    raw_name: 'Ticket Formz',
    display_name: 'Ticket Forms'
  }
  const ticketFields = [
    { id: 1, type: 'text', title_in_portal: 'Description' },
    { id: 2, type: 'subject', title_in_portal: 'Subject' },
    { id: 4, type: 'text', description: 'blah blah', title_in_portal: 'Favorite Burger' },
    { id: 5, type: 'text', title_in_portal: 'Favorite Pizza', required_in_portal: true },
    {
      id: 6,
      type: 'text',
      title_in_portal: 'Favorite Drink',
      editable_in_portal: false,
      visible_in_portal: false
    },
    {
      id: 7,
      type: 'integer',
      title_in_portal: 'Favorite Integer',
      description: 'description number'
    },
    { id: 8, type: 'decimal', title_in_portal: 'Favorite Decimal', required_in_portal: true },
    {
      id: 9,
      type: 'description',
      title_in_portal: 'Favorite Description',
      required_in_portal: true
    },
    {
      id: 10,
      type: 'textarea',
      title_in_portal: 'Favorite Textarea'
    },
    {
      id: 11,
      type: 'checkbox',
      title_in_portal: 'Favorite Checkbox',
      description: 'this is the description',
      required_in_portal: true
    },
    {
      id: 12,
      type: 'tagger',
      title_in_portal: 'Favorite Tagger',
      description: 'this is the tagger description',
      custom_field_options: [
        { name: 'one', value: '1' },
        { name: 'two', value: '2', default: true }
      ],
      required_in_portal: true
    }
  ]
  /* eslint-enable camelcase */

  it('renders the extra fields defined in the ticket form', () => {
    const { container } = renderSubmitTicketForm({
      activeTicketForm,
      ticketFields
    })

    expect(container).toMatchSnapshot()
  })

  it('prefills the form state', () => {
    const setFormState = jest.fn()
    jest.spyOn(i18n, 'getLocale').mockReturnValue('fr')
    renderSubmitTicketForm({
      activeTicketForm,
      ticketFields,
      setFormState,
      ticketFormSettings: [
        { id: 6, prefill: { '*': 'hello world' } },
        { id: 'name', prefill: { '*': 'Lanselot' } },
        { id: 'subject', prefill: { '*': 'Nybeth' } },
        { id: '1', prefill: { fr: 'french', '*': 'fallback' } }
      ],
      ticketFieldSettings: [
        { id: 'email', prefill: { '*': 'blah@blah.com' } },
        { id: 4, prefill: { '*': 'four' } },
        { id: 6, prefill: { '*': 'six' } },
        { id: 5, prefill: { en: 'en', '*': 'fallback', fr: 'fr' } },
        { Terence: 1234 },
        new Date(),
        { Terence: 1234 },
        { id: 1234, Ezel: 'Berbier' }
      ]
    })
    expect(setFormState).toHaveBeenCalledWith(
      expect.objectContaining({
        '6': 'hello world',
        '4': 'four',
        '2': 'Nybeth',
        '5': 'fr',
        '1': 'french',
        name: '', // name cannot be prefilled
        email: '' // email cannot be prefilled
      })
    )
  })
})

describe('preview mode', () => {
  it('disables all fields', () => {
    const { getByLabelText } = renderSubmitTicketForm({
      previewEnabled: true,
      subjectEnabled: true,
      activeTicketForm: null
    })
    expect(getByLabelText('Your name (optional)').disabled).toEqual(true)
    expect(getByLabelText('Email address').disabled).toEqual(true)
    expect(getByLabelText('Subject (optional)').disabled).toEqual(true)
    expect(getByLabelText('How can we help you?').disabled).toEqual(true)
  })
})

describe('form', () => {
  describe('default contact form', () => {
    describe('validation', () => {
      it('displays an error message for name and email fields', () => {
        const submit = jest.fn()
        const { getByText, queryByText } = renderSubmitTicketForm({
          nameFieldRequired: true,
          subjectEnabled: true,
          activeTicketForm: null,
          submit
        })
        expect(queryByText('Please enter a valid name.')).not.toBeInTheDocument()
        expect(queryByText('Please enter a valid email address.')).not.toBeInTheDocument()
        expect(queryByText('Please enter a value.')).not.toBeInTheDocument()

        fireEvent.click(getByText('Send'))

        expect(queryByText('Please enter a valid name.')).toBeInTheDocument()
        expect(queryByText('Please enter a valid email address.')).toBeInTheDocument()
        expect(queryByText('Please enter a value.')).toBeInTheDocument()
        expect(submit).not.toHaveBeenCalled()
      })

      it('displays an error message for invalid email', () => {
        const { getByText, queryByText, getByLabelText } = renderSubmitTicketForm({
          activeTicketForm: null
        })
        fireEvent.change(getByLabelText('Email address'), { target: { value: 'blah' } })
        fireEvent.click(getByText('Send'))

        expect(queryByText('Please enter a valid email address.')).toBeInTheDocument()
      })
    })

    describe('submission', () => {
      it('calls submit handler and changes the button text', () => {
        const submit = jest.fn()
        const { getByText, getByLabelText } = renderSubmitTicketForm({
          nameFieldRequired: true,
          subjectEnabled: true,
          activeTicketForm: null,
          submit
        })
        fireEvent.change(getByLabelText('Email address'), { target: { value: 'hello@world.com' } })
        fireEvent.change(getByLabelText('Your name'), { target: { value: 'Slim Shady' } })
        fireEvent.change(getByLabelText('Subject (optional)'), {
          target: {
            value: "Mom's spaghetti"
          }
        })
        fireEvent.change(getByLabelText('How can we help you?'), { target: { value: 'Help me' } })
        fireEvent.click(getByText('Send'))

        expect(submit).toHaveBeenCalledWith(expect.anything(), {
          isFormValid: true,
          value: {
            description: 'Help me',
            email: 'hello@world.com',
            name: 'Slim Shady',
            subject: "Mom's spaghetti"
          }
        })
        expect(getByText('Submitting...')).toBeInTheDocument()
      })
    })
  })

  describe('ticket forms', () => {
    const activeTicketForm = {
      id: 1,
      raw_name: 'Ticket Formz',
      display_name: 'Ticket Forms'
    }
    const ticketFields = [
      { id: 1, type: 'text', title_in_portal: 'Description', required_in_portal: true },
      { id: 2, type: 'subject', title_in_portal: 'Subject', required_in_portal: true },
      {
        id: 4,
        type: 'text',
        description: 'blah blah',
        title_in_portal: 'Favorite Burger',
        required_in_portal: true
      },
      {
        id: 7,
        type: 'integer',
        title_in_portal: 'Favorite Integer',
        required_in_portal: true
      },
      { id: 8, type: 'decimal', title_in_portal: 'Favorite Decimal', required_in_portal: true },
      {
        id: 10,
        type: 'textarea',
        title_in_portal: 'Favorite Textarea',
        required_in_portal: true
      },
      {
        id: 11,
        type: 'checkbox',
        title_in_portal: 'Favorite Checkbox',
        required_in_portal: true
      },
      {
        id: 12,
        type: 'tagger',
        title_in_portal: 'Favorite Tagger',
        custom_field_options: [
          {
            name: 'one',
            value: 'one'
          },
          {
            name: 'two',
            value: 'two'
          },
          {
            name: 'three',
            value: 'three'
          }
        ],
        required_in_portal: true
      }
    ]

    describe('validation', () => {
      it('renders the error messages', () => {
        const { container, getByText } = renderSubmitTicketForm({
          activeTicketForm,
          ticketFields
        })
        fireEvent.click(getByText('Send'))
        expect(container).toMatchSnapshot()
      })

      it('validates number fields', () => {
        const { getAllByText, getByText, getByLabelText } = renderSubmitTicketForm({
          activeTicketForm,
          ticketFields
        })
        fireEvent.change(getByLabelText('Favorite Integer'), { target: { value: 'hello' } })
        fireEvent.change(getByLabelText('Favorite Decimal'), { target: { value: 'world' } })
        fireEvent.click(getByText('Send'))
        expect(getAllByText('Please enter a number.').length).toEqual(2)
      })
    })

    const renderWithState = (state, renderer) => {
      const submit = jest.fn()
      const setFormState = jest.fn()
      const getLatestState = () => {
        const calls = setFormState.mock.calls
        return calls[calls.length - 1][0]
      }
      return {
        submit,
        getLatestState,
        ...renderSubmitTicketForm(
          {
            formState: state,
            activeTicketForm,
            ticketFields,
            submit,
            setFormState
          },
          renderer
        )
      }
    }

    const ticketFieldAct = (state, renderer, cb) => {
      const getState = renderWithState(state, renderer).getLatestState
      cb()
      return getState()
    }

    describe('submission', () => {
      it('submits the expected object and changes the button text', () => {
        let state = {}
        const { container, rerender, getByText, getByLabelText } = renderWithState(state)
        fireEvent.click(container.querySelector('[data-garden-id="dropdowns.select"]'))
        fireEvent.click(getByText('three'))
        jest.runAllTimers()
        fireEvent.change(getByLabelText('Email address'), { target: { value: 'hello@world.com' } })
        fireEvent.change(getByLabelText('Your name (optional)'), {
          target: { value: 'Slim Shady' }
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.change(getByLabelText('Favorite Integer'), { target: { value: '99' } })
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.change(getByLabelText('Favorite Decimal'), { target: { value: '100.00' } })
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.change(getByLabelText('Description'), { target: { value: 'despacito' } })
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.change(getByLabelText('Subject'), { target: { value: 'subject' } })
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.change(getByLabelText('Favorite Textarea'), { target: { value: 'textarea' } })
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.change(getByLabelText('Favorite Burger'), { target: { value: 'burger' } })
        })
        state = ticketFieldAct(state, rerender, () => {
          fireEvent.click(getByLabelText('Favorite Checkbox'))
        })
        const { submit } = renderWithState(state, rerender)
        fireEvent.click(getByText('Send'))
        expect(submit).toHaveBeenCalledWith(expect.anything(), {
          isFormValid: true,
          value: {
            email: 'hello@world.com',
            name: 'Slim Shady',
            '1': 'despacito',
            '10': 'textarea',
            '11': 1,
            '12': 'three',
            '2': 'subject',
            '4': 'burger',
            '7': '99',
            '8': '100.00'
          }
        })
        expect(getByText('Submitting...')).toBeInTheDocument()
      })
    })
  })
})
