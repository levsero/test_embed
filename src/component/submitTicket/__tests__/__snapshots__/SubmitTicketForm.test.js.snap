// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#renderNameField renders the name field with correct name props 1`] = `
<input
  aria-invalid="false"
  aria-labelledby="garden-text-field-0--label"
  class="Input-sc-4gf5js-0 text__c-txt__input___2ZeOl kdlara"
  data-garden-id="textfields.input"
  data-garden-version="3.7.0"
  id="garden-text-field-0--input"
  name="name"
  value=""
/>
`;

exports[`#renderNameField renders the name field with correct name props when it is required 1`] = `
<input
  aria-invalid="false"
  aria-labelledby="garden-text-field-4--label"
  class="Input-sc-4gf5js-0 text__c-txt__input___2ZeOl kdlara"
  data-garden-id="textfields.input"
  data-garden-version="3.7.0"
  id="garden-text-field-4--input"
  name="name"
  required=""
  value=""
/>
`;
